### ISSUE TITLE
Support Wayland

### ISSUE THREAD

**Issue by liZe** [fontforge/fontforge#3295]
Since the GDK backend has been merged (see #2782), it is now possible to launch Fontforge as a real Wayland application with no X server.

Here's what is working now using the Wayland GDK3 backend:
- [x] Application starts
- [x] Dialog windows are drawn correctly and usable
- [x] Menus, buttons, almost everything UI-related is rendered correctly

What needs to be fixed:
- [] Missing client-side decorations (known issue),
- [] Some UI interactions are buggy, for example when clicking on menus while moving the mouse (probably related to some missing focus-related stuff from the window manager),
- [] Everything that displays a character, ie the main window and the metrics window, are white (the characters are never drawn),
- [] The character editor, including the buttons, is displayed once, but never re-drawn when modified.

### When reporting a bug/issue:
- [] Screenshot
- [x] The FontForge version and the operating system you're using
Git version 88c13bdee with GDK3 backend
- [x] The behavior you expect to see, and the actual behavior
Everything should work as it works on X
- [x] Steps to reproduce the behavior
Launch Fontforge on Wayland
- [x] \(optional) Possible solution/fix/workaround
Looks like the drawing areas are not refreshed, I'll check the current code to understand how it works and how to fix it. If anyone has an idea about what's going on, I'd love to get some hints!

I'll add screenshots too if anyone needs them!

**Comment by jtanx** [comment]
So I've been recently looking at the GTK port again and come to the following conclusions:

* Wayland makes it near-impossible to do some things required by gdraw. Most notably, setting the position of windows and translating coordinates between unrelated windows is effectively not supported. This mostly affects popups and menus.
* Migration of GDraw to GTK4/GDK4 will likely never occur as a result, since GTK/GDK4 reforms the API to be more wayland-centric, deleting a bunch of functions that were needed (the ability to move and resize windows, and also the whole concept of root coordinates)
* Migration to GTK3 is also somewhat questionable and without reworking GDraw in some fashion, will still be broken by the aforementioned restrictions.
* The cost of trying to rework GDraw to fit in this may not be worth it vs. rewriting to just not use GDraw instead.

In essence I think we're going to be stuck with having to use XWayland for a long time yet.

**Comment by skef** [comment]
Not entirely on topic, but ...

I've become pessimistic about there being a future for FontForge that's much different than its present because of the combination:

1. The UI stuff is too low level to work with productively.
2. Porting the whole UI to a different layer seems like too great a task given all the bitmap font stuff that most project members don't look at much anymore. 
3. A "slim FontForge" without the bitmap stuff would probably suffer the same fate as all the other forks.

If there's a way around this it seems like the way to go would be to start cleaning up the internals and make a clearer abstraction layer separating them from the UI so that you could build up a new GUI with slimmed down functionality that could coexist with "FontForge" as it's broadly understood. OTOH it's not obvious that the source base combined with the usual level of participation would make that viable, or more viable than something like the RuneBender project. Most people don't want to deal with code written the way FontForge is and rewriting it (into C++ classes or just with safer C macros or inlines) would be a huge project.

**Comment by jtanx** [comment]
I tend to agree with you. I doubt there will be any major changes in this space any time soon, unfortunately. But hey, if I end up with a bunch of free time ever again I wouldn't mind tackling it 😄

**Comment by ctrlcctrlv** [comment]
I unfortunately agree. I've been working on my own tools. FontForge is basically just my glyph editor these days, I use fontmake for all my important fonts. Though I did play with it a bit for my toy font FRB Cistercian. I found bugs by doing that, but not any I feel like tackling.

**Comment by jtanx** [comment]
I've recently been playing with porting to qt, and it's actually been quite good so far - I know I've mentioned this possibility before but I wish I tried sooner.

Everything I've done so far more or less 'just works'; Qt even figured out how to do all the things I mentioned were impossible with GTK, which is nice. So this is actually a real possibility now.

If I can find the time (ha), I'd want to pursue this quite seriously, as programming for Qt is just leaps and bounds better than working with the broken mess of GTK

**Comment by lilydjwg** [comment]
GTK (or to be specific, pango) has a feature to render characters absent in fonts as a square with hex code point numbers in it. Can you do that in the Qt version?

**Comment by ctrlcctrlv** [comment]
Hmm. @jtanx I think that your idea of porting to Qt is great for more than one reason but it will cause you to have to convert the code base to C++.

I think that this is overall positive because it will allow us to use many useful constructs and types in that language. However I would like to see you do some separation between which parts will be C++ and which not.

I would say that all of the functions that are accessible from Python without using the UI should be extern C and not converted to C++. If it is not done this way I'm not going to be able to link to it as easily and that's going to be a problem for me.

I know that linking is not officially supported of course. However as a courtesy I think it should be kept until you decide that the function individually should be converted to C++ because there is some construct that makes so much sense and will fix a bug and make the code better.

Your thoughts on this?

**Comment by jtanx** [comment]
Right now, I'm only implementing the gdraw interface, so yes I'm already constraining where I use C++ to just that most low-level layer.

I would suspect the majority of C++ code in the future would be GUI related anyway which doesn't even end up in the dll; as to what happens down the line for doing anything else in C++, if ever, we can cross that bridge when we come to it

**Comment by jtanx** [comment]
> GTK (or to be specific, pango) has a feature to render characters absent in fonts as a square with hex code point numbers in it. Can you do that in the Qt version?

If you care about such things you can do anything

> Another recent pango feature is that, it auto selects an emoji font for emoji characters. Qt itself doesn't do that too. (And preferring emoji fonts for sans-serif in fontconfig has some issues.)

Yea this actually works better in qt
![image](https://user-images.githubusercontent.com/5137410/155488611-738ef30b-5571-4833-b54a-0cf08cd7c573.png)

 than current gdk
![image](https://user-images.githubusercontent.com/5137410/155488698-9c05fdee-38e6-46dc-be2b-a631378e26ec.png)

**Comment by iorsh** [comment]
Tried to run FontForge on Kubuntu 22.04 with Wayland, and somehow it just worked and looked pretty usual. Could it be that the problem resolved itself, or maybe I'm missing something?

**Comment by zefr0x** [comment]
> Tried to run FontForge on Kubuntu 22.04 with Wayland, and somehow it just worked and looked pretty usual. Could it be that the problem resolved itself, or maybe I'm missing something?

You are using XWayland. You can check using `xorg-xeyes`.

**Comment by Integral-Tech** [comment]
Any progress on this issue? Both KDE and GNOME are taking steps towards dropping X11.

**Comment by iorsh** [comment]
> Any progress on this issue? Both KDE and GNOME are taking steps towards dropping X11.

I just merged https://github.com/fontforge/fontforge/pull/5546, which kicks off the transition to GTK3 UI toolkit. While I don't expect it to happen quickly, this is a major first step, and at least some Wayland issues would be gradually resolved as we move.
